Library/Homebrew/test/utils/git_spec.rb

```diff
@@ -0,0 +1,231 @@
+      described_class.send(:remove_instance_variable, :@git)
+    end
+  end
+
+  describe "::git_available?" do
+    it "returns true if git --version command succeeds" do
+      allow_any_instance_of(Process::Status).to receive(:success?).and_return(true)
+      expect(described_class.git_available?).to be_truthy
+    end
+
+    it "returns false if git --version command does not succeed" do
+      allow_any_instance_of(Process::Status).to receive(:success?).and_return(false)
+      expect(described_class.git_available?).to be_falsey
+    end
+
+    it "returns git version if already set" do
+      described_class.instance_variable_set(:@git, true)
+      expect(described_class.git_available?).to be_truthy
+      described_class.instance_variable_set(:@git, nil)
+    end
+  end
+
+  describe "::git_path" do
+    context "when git is not available" do
+      before do
+        described_class.instance_variable_set(:@git, false)
+      end
+
+      it "returns nil" do
+        expect(described_class.git_path).to eq(nil)
+      end
+    end
+
+    context "when git is available" do
+      before do
+        described_class.instance_variable_set(:@git, true)
+      end
+
+      it "returns path of git" do
+        expect(described_class.git_path).to end_with("git")
+      end
+
+      it "returns git_path if already set" do
+        described_class.instance_variable_set(:@git_path, "git")
+        expect(described_class.git_path).to eq("git")
+        described_class.instance_variable_set(:@git_path, nil)
+      end
+    end
+  end
+
+  describe "::git_version" do
+    context "when git is not available" do
+      before do
+        described_class.instance_variable_set(:@git, false)
+      end
+
+      it "returns nil" do
+        expect(described_class.git_path).to eq(nil)
+      end
+    end
+
+    context "when git is available" do
+      before do
+        described_class.instance_variable_set(:@git, true)
+      end
+
+      it "returns version of git" do
+        expect(described_class.git_version).not_to be_nil
+      end
+
+      it "returns git_version if already set" do
+        described_class.instance_variable_set(:@git_version, "v1")
+        expect(described_class.git_version).to eq("v1")
+        described_class.instance_variable_set(:@git_version, nil)
+      end
+    end
+  end
+
+  describe "::ensure_git_installed!" do
+    it "returns nil if git already available" do
+      described_class.instance_variable_set(:@git, true)
+      expect(described_class.ensure_git_installed!).to be_nil
+    end
+
+    it "can't install brewed git if homebrew/core is unavailable" do
+      described_class.instance_variable_set(:@git, false)
+      allow_any_instance_of(Pathname).to receive(:directory?).and_return(false)
+      expect { described_class.ensure_git_installed! }.to raise_error("Git is unavailable")
+    end
+
+    it "installs git if homebrew/core is available and git is not available already" do
+      # WIP
```

Comment:
If I mock `git_available?` to return false and also mock the existence of `CoreTap.instance.formula`, running `safe_system` to install git raises an error since git is already installed on my system. What could be the workaround to mock a system that doesn't have brew installed git?
Thanks!
@mistydemeo @MikeMcQuaid @reitermarkus